### ISSUE TITLE
"Multihash too short" error in storetheindex while the ad is correct in index-provider

### ISSUE THREAD

**Issue by ischasny** [filecoin-project/storetheindex#940]
Recently storetheindex has had a few `Multihash too short` errors while processing the ad chain for `12D3KooWRNijznEQoXrxBeNLb2TqbSFm8gG8jKtfEsbC1C9nPqce` (Collab cluster). The log can be seen [here](https://protocollabs.grafana.net/explore?orgId=1&left=%7B%22datasource%22:%22grafanacloud-logs%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22grafanacloud-logs%22%7D,%22expr%22:%22%7Bnamespace%3D%5C%22storetheindex%5C%22%7D%20%22,%22refId%22:%22A%22,%22editorMode%22:%22code%22,%22queryType%22:%22range%22%7D,%7B%22refId%22:%22B%22,%22editorMode%22:%22builder%22,%22expr%22:%22%7Bcontainer_name%3D%5C%22indexer%5C%22,%20environment%3D%5C%22dev%5C%22%7D%20%7C%3D%20%60baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihsxwd6wy2a%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22grafanacloud-logs%22%7D%7D%5D,%22range%22:%7B%22from%22:%221665874800000%22,%22to%22:%221665910800000%22%7D%7D).

After having closer look, all multihashes in the ad were valid on the index provider side:
[baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihsxwd6wy2a.txt](https://github.com/filecoin-project/storetheindex/files/9800648/baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihsxwd6wy2a.txt)

Snippet of code to verify:
```
	file, err := os.Open("/baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihsxwd6wy2a.txt))
        require.NoError(t, err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := strings.TrimSpace(scanner.Text())
		mh, err := multihash.FromB58String(text)
		require.NoError(t, err)
	}
        require.NoError(t, scanner.Err())
```

1.  A short term fix: if a multihash is invalid - continue ingestion without bailing early. Otherwise an invalid multihash could prevent the entire chain from being processed.
2. Investigate a bug. Could that be specific to a datastore type?

**Comment by gammazero** [comment]
This provider is only listed in the providers for `dev.cid.contact` and not for `cid.contact`:
```
{"AddrInfo":{"ID":"12D3KooWRNijznEQoXrxBeNLb2TqbSFm8gG8jKtfEsbC1C9nPqce","Addrs":["/ip4/147.75.87.211/tcp/4001","/ip4/147.75.87.211/tcp/4002/ws","/ip4/147.75.87.211/udp/4001/quic","/ip6/2604:1380:4601:f600::3/tcp/4001","/ip6/2604:1380:4601:f600::3/tcp/4002/ws","/ip6/2604:1380:4601:f600::3/udp/4001/quic"]},"LastAdvertisement":{"/":"baguqeeras525cdadx5yvqxayuy2tmpl7zvi5k4muurww72swjcdwskxcgpca"},"LastAdvertisementTime":"2022-10-16T06:04:08Z","Publisher":{"ID":"12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU","Addrs":["/ip4/147.75.87.211/tcp/3103"]}}
```

The publisher is `2D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU`.  Sync with that publisher:
```
./storetheindex admin sync --pubid=12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU --addr=/ip4/147.75.87.211/tcp/3103
```
Results in this add being ingested correctly:
```
2022-10-17T17:02:58.513Z    INFO    indexer/ingest  ingest/ingest.go:956    Processing advertisement    {"adCid": "baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihsxwd6wy2a", "entriesCid": "baguqeerasnhl6dwdwhivsek6tbrz36jeml4ycqlqrnbldnnfmx\
4unmehrega", "publisher": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", "progress": "609 of 3150"}                                                                                                                                          
2022-10-17T17:02:58.515Z    INFO    go-legs go-legs@v0.4.16/subscriber.go:505   Start sync  {"peer": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", "cid": "baguqeerasnhl6dwdwhivsek6tbrz36jeml4ycqlqrnbldnnfmx4unmehrega"}                  
2022-10-17T17:02:58.685Z    INFO    go-legs-dtsync  dtsync/sync.go:251  datatransfer completed successfully {"cid": "baguqeerasnhl6dwdwhivsek6tbrz36jeml4ycqlqrnbldnnfmx4unmehrega", "peer": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU"}  
2022-10-17T17:02:58.687Z    INFO    go-legs go-legs@v0.4.16/subscriber.go:1015  Sync completed  {"cid": "baguqeerasnhl6dwdwhivsek6tbrz36jeml4ycqlqrnbldnnfmx4unmehrega", "peer": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU"}              
2022-10-17T17:02:58.748Z    INFO    indexer/ingest  ingest/linksystem.go:536    Put multihashes in entry chunk  {"publisher": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", "adCid": "baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihs\
xwd6wy2a", "contextID": "mgQ0J2srPL071/EJQfjkQCOgNk0CgzwBJFXkuilhjnw=", "provider": "12D3KooWRNijznEQoXrxBeNLb2TqbSFm8gG8jKtfEsbC1C9nPqce", "entriesKind": "EntryChunk", "count": 3616}                                                               
2022-10-17T17:02:58.750Z    INFO    go-legs go-legs@v0.4.16/subscriber.go:505   Start sync  {"peer": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", "cid": "baguqeeral24kg3uqkm3szpnzejvysqaffxspsfjehzk5rhv652mdfwmanzsa"}                  
2022-10-17T17:02:59.334Z    INFO    indexer/ingest  ingest/linksystem.go:536    Put multihashes in entry chunk  {"publisher": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", "adCid": "baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihs\
xwd6wy2a", "contextID": "mgQ0J2srPL071/EJQfjkQCOgNk0CgzwBJFXkuilhjnw=", "provider": "12D3KooWRNijznEQoXrxBeNLb2TqbSFm8gG8jKtfEsbC1C9nPqce", "entriesKind": "EntryChunk", "count": 16384}                                                              
2022-10-17T17:02:59.338Z    INFO    go-legs-dtsync  dtsync/sync.go:251  datatransfer completed successfully {"cid": "baguqeeral24kg3uqkm3szpnzejvysqaffxspsfjehzk5rhv652mdfwmanzsa", "peer": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU"}  
2022-10-17T17:02:59.338Z    INFO    go-legs go-legs@v0.4.16/subscriber.go:1070  Segmented sync completed    {"cid": "baguqeeral24kg3uqkm3szpnzejvysqaffxspsfjehzk5rhv652mdfwmanzsa", "peer": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", \
"syncedCidCount": 1}                                                                                                                                                                                                                                  
2022-10-17T17:02:59.338Z    INFO    indexer/ingest  ingest/linksystem.go:158    Finished syncing advertisement  {"publisher": "12D3KooWPVeKXN34kS5YLGho1L3SFkcTsAPfmxXNYpYwpf1yMQrU", "adCid": "baguqeera373ixoiaxibql54mmcqc57roe2k6mbjhj4wowbiogihs\
xwd6wy2a", "contextID": "mgQ0J2srPL071/EJQfjkQCOgNk0CgzwBJFXkuilhjnw=", "provider": "12D3KooWRNijznEQoXrxBeNLb2TqbSFm8gG8jKtfEsbC1C9nPqce", "entriesKind": "EntryChunk", "elapsed": "824.986308ms", "multihashes": 20000}                             
```

**Comment by ischasny** [comment]
Thanks @gammazero. Is that after the fix?

**Comment by gammazero** [comment]
The entries were delivered in 2 chunks, the first received/processed contained 3616 multihashes. So, the failure is on the first chunk, [here](https://github.com/filecoin-project/storetheindex/blob/main/internal/ingest/linksystem.go#L400).

Execution continues [here](https://github.com/filecoin-project/storetheindex/blob/main/internal/ingest/linksystem.go#L474) and then [here](https://github.com/filecoin-project/storetheindex/blob/main/internal/ingest/linksystem.go#L533), then into go-indexer-core [here](https://github.com/filecoin-project/go-indexer-core/blob/main/engine/engine.go#L128), then [here](https://github.com/filecoin-project/go-indexer-core/blob/main/store/storethehash/storethehash.go#L102).

This then calls into the value store implementation, [here](https://github.com/filecoin-project/go-indexer-core/blob/e0dd6fbb6780156438ab72a3ef7589df602d63a3/store/storethehash/storethehash.go#L430), which then tries to decode the key for the multihash primary storage [here](https://github.com/ipld/go-storethehash/blob/master/store/store.go#L315), which then tries to decode the key as a multihash [here](https://github.com/ipld/go-storethehash/blob/0da08aea7d1bec8c16c881eceb980d84c1525c46/store/primary/multihash/multihash.go#L335), which then gets the error decoding the multihash [here](https://github.com/multiformats/go-multihash/blob/master/multihash.go#L225) when reading data [here](https://github.com/multiformats/go-multihash/blob/master/multihash.go#L270-L271)

**Comment by gammazero** [comment]
This is fixed in https://github.com/ipld/go-storethehash/pull/82

There is no need to apply any additional filtering because bad multihashes are now caught here: https://github.com/filecoin-project/storetheindex/blob/main/internal/ingest/linksystem.go#L499

**Comment by gammazero** [comment]
Fixed by #942
